src/main/java/com/aws/greengrass/util/SdkClientWrapper.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.util;
+
+import software.amazon.awssdk.core.SdkClient;
+import software.amazon.awssdk.core.exception.SdkClientException;
+
+import java.net.SocketException;
+import java.util.concurrent.locks.Lock;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public final class SdkClientWrapper<T extends SdkClient> {
+    private volatile T client;
+    private final Supplier<T> clientFactory;
+    private final Lock lock = LockFactory.newReentrantLock(this);
+
+    public SdkClientWrapper(Supplier<T> clientFactory) {
+        this.clientFactory = clientFactory;
+        this.client = clientFactory.get();
+    }
+
+    public <R> R execute(final Function<T, R> operation) {
+        try {
+            return executeOperation(operation);
+        } catch (Exception e) {
+            if (shouldRefreshClient(e)) {
+                System.out.println("Client needs refresh due to: " + e.getMessage());
+                refreshClient();
+                return executeOperation(operation);
+            }
+            throw wrapException(e, "SDK operation failed");
+        }
+    }
+
+    private <R> R  executeOperation(final Function<T, R> operation) {
+        try {
+            return operation.apply(client);
+        } catch (Exception e) {
+            throw wrapException(e, "SDK operation failed");
+        }
+    }
+
+    private RuntimeException wrapException(Exception e, String message) {
+        if (e instanceof RuntimeException) {
+            return (RuntimeException) e;
+        }
+        return SdkClientException.builder()
+                .message(message + ": " + e.getMessage())
+                .cause(e)
+                .build();
+    }
+
+    private void refreshClient() {
+        synchronized (lock) {
+            if (client != null) {
+                try {
+                    client.close();
+                } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It appears that your code handles a broad swath of exceptions in the catch block, potentially trapping dissimilar issues or problems that should not be dealt with at this point in the program.

[Learn more](https://cwe.mitre.org/data/definitions/396.html)